Migrate lib to coroutines and remove RX dependencies (#20)
hyperdevs-team/mini-kotlin: README.md

# Mini
[![Release](https://jitpack.io/v/hyperdevs-team/mini-kotlin.svg)](https://jitpack.io/#hyperdevs-team/mini-kotlin)

> ## ⚠️ DOCS HAVE TO BE UPDATED FOR VERSION 2.x.y

Reviewer: ```suggestion
## ⚠️ DOCS HAVE TO BE UPDATED FOR VERSION 3.x.y
```

Author: Done in 839877ca927656bc40554e6dea7cbf83e22bdac1

> ## ⚠️ DOCS HAVE TO BE UPDATED FOR VERSION 3.x.y

Reviewer: I have added some examples of use of the Kodein utils in general, not just of the new part, as we were lacking of them. Check if you miss something.
63e044a01e8f169b2add9ea95905f36e99de4207

```groovy
dependencies {
    def mini_version = "3.0.0"
    // Minimum working dependencies
    implementation "com.github.hyperdevs-team.mini-kotlin:mini-android:$mini_version"
    kapt "com.github.hyperdevs-team.mini-kotlin:mini-processor:$mini_version"

    // Kodein helper libraries
    implementation "com.github.hyperdevs-team.mini-kotlin:mini-kodein:$mini_version"

>     implementation "com.github.hyperdevs-team.mini-kotlin:mini-kodein-android:$mini_version"

Reviewer: 63e044a01e8f169b2add9ea95905f36e99de4207

```groovy
dependencies {

>     def mini_version = "3.0.0"

Reviewer 2: Remove RX dependencies

    store.initialize()
}

// Optional: add logging middleware to log action events
dispatcher.addMiddleware(LoggerMiddleware(stores)) { tag, msg ->
    Log.d(tag, msg)
}
```

### \[Android] Kodein Android utils

> `mini-kodein-android` has some utils methods in order to inject an Android's `ViewModel` in a `DIAware` `Activity` or `Fragment`.

Reviewer 3: ```suggestion
`mini-kodein-android` has some utility methods in order to inject an Android's `ViewModel` in a `DIAware` `Activity` or `Fragment`.
```

> To use it, bind with kodein the Android's `ViewModelProvider.Factory` instance:

Reviewer 3: ```suggestion
In order to use them, bind the Android's `ViewModelProvider.Factory` instance with Kodein:
```

```kotlin
// Use any tag to differ between the injected `Context` or `Application` if you are binding also `Context` with Kodein
bind<Application>("appTag") with singleton { app }
bind<ViewModelProvider.Factory>() with singleton { DIViewModelFactory(di.direct) }
```

> To inject a view model without parameters, bind it as follows in this example:

Reviewer 3: ```suggestion
To inject a ViewModel without parameters, bind it as follows:
```

```kotlin
bindViewModel { MainViewModel(instance("appTag") }
```
And in your `DIAware` `Activity` or `Fragment`:
```kotlin
private val mainViewModel: MainViewModel by viewModel()
```

> `mini-kodein-android-compose` has some utils methods in order to inject an Android's `ViewModel` in the scope of a Navigation component

Reviewer 3: ```suggestion
`mini-kodein-android-compose` has some utility methods in order to inject an Android's `ViewModel` in the scope of a Navigation component
```

> graph. This is useful as in Jetpack Compose is common to have only one or few `Activities` and none `Fragment` so, in order to scope

Reviewer 3: ```suggestion
graph. This is useful as in Jetpack Compose it is common to have only one or few `Activities` and no `Fragment`s so, in order to scope
```

the lifecycle of the `ViewModel` not for all the life of the `Activity`, we can scope it to any route existing in the `NavBackStackEntry`.

> For use it, do the same as above, but instead of injecting the view model scoped to a route of the Navigation, the `NavHost` composable

Reviewer 3: ```suggestion
In order to use it, do the same as above, but instead of injecting the ViewModel scoped to a route of the Navigation, the `NavHost` composable
```

must be inside an `DIAware Activity`, and then do as follows:
```kotlin
composable(route = "home") { navBackStackEntry ->
    val homeViewModel: HomeViewModel by navBackStackEntry.viewModel(contextDI())
    HomeScreen(homeViewModel, ...)
}
```

> In case you want to pass an argument to a view model, you need to bind the factory of that kind of Android's ViewModel.

Reviewer 3: ```suggestion
In case you want to pass an argument to a ViewModel, you need to bind the factory of that kind of Android's ViewModel.
```

You can do this in both `mini-kodein-android`, and `mini-kodein-android-compose`.

> For example given a view model that you want to pass a `String` param, it would be:

Reviewer 3: ```suggestion
For example, given a `ViewModel` that you want to pass a `String` param, it would be:
```

```kotlin
bindViewModelFactory<HomeViewModelWithParameter, ViewModelProvider.Factory> { param ->
    TypedViewModelFactory(HomeViewModelWithParameter::class, instance("appTag"), param as String)
}
```

> And for retrieve it with the given param in its constructor:

Reviewer 3: ```suggestion
And to retrieve it with the given param in its constructor:
```
